rfcs/0005-pablo-distribution.md

```diff
@@ -0,0 +1,667 @@
+        pub reward_epoch_start: Epoch,
+        /// List of reward asset/pending rewards.
+        pub pending_rewards: Rewards,
+        /// The date at which this NFT was minted.
+        pub lock_date: Timestamp,
+        /// The duration for which this NFT stake was locked.
+        pub lock_duration: DurationSeconds,
+        /// The penalty applied if a staker unstake before the end date.
+        pub early_unstake_penalty: Penalty<AccountId>,
+        /// The reward multiplier.
+        pub reward_multiplier: Perbill,
+    }
+
+This data structure can be modified to support the RP approach by
+adding/adjusting/removing some fields.
+
+    /// This can possibly be renamed as `RewardPoolPosition`
+    pub struct StakingNFT<RewardPoolId, AccountId, AssetId, Balance> {
+        /// [Adding] Reward Pool ID from which pool to allocate rewards for this
+        pub reward_pool_id: RewardPoolId,
+
+        /// The staked asset.
+        pub asset: AssetId,
+        /// The original stake this NFT was minted for.
+        /// Used for calculating the pool share.
+        pub stake: Balance,
+
+        /// [Adding] Pool share received for this NFT
+        pub share: Balance,
+
+        /// [Adding] reduced rewards by asset for the fNFT (d_n)
+        pub reductions: BtreeMap<AssetId, Balance>,
+
+        /// [Removing] The reward epoch at which this NFT will start yielding rewards. This is no longer needed.
+        /// pub reward_epoch_start: Epoch,
+
+        /// [Removing] List of reward asset/pending rewards. This is no longer needed as pending rewards are pooled
+        /// pub pending_rewards: Rewards,
+
+        /// The date at which this NFT was minted.
+        pub lock_date: Timestamp,
+
+        /// The duration for which this NFT stake was locked.
+        pub lock_duration: DurationSeconds,
+
+        /// The penalty applied if a staker unstake before the end date.
+        pub early_unstake_penalty: Penalty<AccountId>,
+
+        /// [Removing] The reward multiplier. Can remove this and calculate the extra shares a user gets based on the multiplier
+        /// pub reward_multiplier: Perbill,
+    }
+
+Now in order to allow redeeming the above fNFT, following data
+structures needs to be tracked in the staking rewards pallet,
+
+    /// Total reward for a particular asset
+    pub struct Reward<AssetId, Balance, BlockNumber> {
+
+        pub asset_id: AssetId,
+
+        /// Total rewards including inflation for adjusting for new stakers joining the pool. All stakers
+        /// in a pool are eligible to receive a part of this value based on their share of the pool.
+        pub total_rewards: Balance,
+
+        /// A book keeping field to track the actual total reward without the reward inflation caused
+        /// by new stakers joining the pool.
+        pub total_inflation: Balance,
+
+        /// Upper bound on the `actual_total_rewards`.
+        pub max_rewards: Balance,
```

Comment:
i do not agree with this implementaion. max_rewards can be defined in terms of rate and time of rate active.